File: Daily_Challenge/20230328_LC_1402.py
class Solution:
    def maxSatisfaction(self, satisfaction: List[int]) -> int:
        # this is too smart....
        res = total = 0
        A.sort()
        while A and A[-1] + total > 0:
            total += A.pop()
            res += total
        return res
        '''
        satisfaction.sort()
        n = len(satisfaction)
        mem = {}

        def dp(i, t):
            if i >= n:
                return 0
            if (i, t) in mem:
                return mem[(i, t)]
            mem[(i, t)] = 0
            # choose to cook
            mem[(i, t)] = max(mem[(i, t)], t * satisfaction[i] + dp(i+1, t+1))
            # choose to not cook
            mem[(i, t)] = max(mem[(i, t)], dp(i+1, t))
            return mem[(i, t)]
        
        return dp(0, 1)
        '''

File: Daily_Challenge/20230328_LC_1402.cpp
class Solution {
public:
    int dp(int i, int t, unordered_map<int, unordered_map<int, int>>& mem, vector<int>& satisfaction){
         if (i >= satisfaction.size()){
             return 0;
         }
         if (mem.count(i) && mem[i].count(t)){
             return mem[i][t];
         }
         mem[i][t] = 0;
         // choose to cook
         mem[i][t] = max(mem[i][t], t * satisfaction[i] + dp(i+1, t+1, mem, satisfaction));
         // choose to not cook
         mem[i][t] = max(mem[i][t], dp(i+1, t, mem, satisfaction));
         return mem[i][t];
    }
    int maxSatisfaction(vector<int>& satisfaction) {
        sort(satisfaction.begin(), satisfaction.end());
        unordered_map<int, unordered_map<int, int>> mem;

        return dp(0, 1, mem, satisfaction);
    }
};